rfcs/0000-ret-cont-recursive-nix.md

```diff
@@ -0,0 +1,250 @@
+---
+feature: ret-cont-recursive-nix
+start-date: 2019-02-01
+author: John Ericson (@Ericson2314)
+co-authors: (find a buddy later to help our with the RFC)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+"Ret-cont" recursive Nix is a restricted form of recursive Nix, where one builds a derivations instead of executing builds during the builds.
+This avoids some platform-specific contortions relating to nested sandboxing.
+More importantly, it prevents imperative and overly linear direct-style build scripts;
+easy to write but throwing away the benefits of Nix.
+
+# Motivation
+[motivation]: #motivation
+
+The benefits of recursive Nix have been described in many places.
+One main reason is if we want Nix to function as a build system and package manager, we need upstream packages to use Nix too without duplicating their build systems in Nixpkgs.
+For this case, people usually imagine derivations like
+```nix
+{ stdenv, pkgs, nix }:
+
+stdenv.mkDerivation {
+  name = "foo";
+  version = "1.2.3";
+
+  src = ...;
+
+  nativeBuildInputs = [ nix ];
+  NIX_PATH = "nixpkgs=${pkgs.path}";
+
+  outputs = [ "out" "dev" ];
+
+  doConfigure = false;
+  doBuild = false;
+
+  installPhase = ''
+    for o in $outputs; do
+      pkg=$(nix-build -E '((import <nixpkgs> {}).callPackage ./. {}).'"$o")
+      cp -r $pkg ${!o}
+    done
+  '';
+}
+```
+The other main reason is other build systems should be translated to Nix without vendoring tons of autogenerated code in Nixpkgs.
+For this, case, the one difference is we need to generate some Nix first.
+```nix
+stdenv.mkDerivation {
+  # ...
+  installPhase = ''
+    bazel2nix # new bit
+    for o in $outputs; do
+      pkg=$(nix-build -E '((import <nixpkgs> {}).callPackage ./. {}).'"$o")
+      cp -r $pkg ${!o}
+    done
+  '';
+}
+```
+
+"Ret-cont" recursive Nix, short for "return-continuation" recursive Nix, is a different take on recursive Nix.
+The normal variant in the examples above might be termed "direct-style" recursive Nix.
+Consider what happens with the recursive "nix-build" in those examples:
+the outer build blocks while the inner one builds, and then the other one continues.
+Just as we can CPS-transform programs, reifying the context of a function call as another function (which is passed as an argument), so we can imagine splitting the derivation in two at this blocking point.
+This gives the "continuation" part of the name.
+But whereas the CPS transformation makes the continuation an argument, the Nix *derivation* language is first order.
+Instead, we can produce a derivation which has the callee as a dependency, and continuation drv downstream depending on it.
+Since the outer derivation evaluates (builds) the inner derivation rather than calling anything, I deem that it returns the derivation.
+This gives the "return" part of the name.
+Both differences together, the first example becomes something like:
```

Comment:
But the first example had two outputs - this one has only one. Does this mean I can only recurse once?

Reply:
You can recurse as many times as you want. That one output is the derivation itself, not some output of it. After you've been rewritten into the final derivation (i.e. a non-recursive one), you produce all of it's outputs.

Reply:
Ok, I suppose my question more meant: can you provide behaviour equivalent to your first example. If so, could that be demonstrated in this RFC?

Reply:
The later ones are equivalent. Perhaps my typo I fixed in https://github.com/NixOS/rfcs/pull/40/commits/5e56f21125faa6e92ff8269e8fd35655ad9ab958 was causing confusion?